### PR TITLE
Governance changes: clearer timelines and conditions for decision making

### PR DESCRIPTION
**Pull request by fkiraly**
For consideragin in the 2021-07-05 CC meeting, I've drafted proposed changes to the governance guidelines that aim to make decision conditions for PR acceptance clearer and put clear timelines on decisions.

Major changes:
* making concrete conditions for core developers becoming active or inactive
* making explicit that core developer approval for proposals must be by a core developer who is not the proposer
* added lazy lazy approval for proposals made by core devs (new), so core dev PRs do not indefinitely lie around without approval - this is based on assumption that core devs know what they are doing
* clarified where approvals/rejections must be put
* replaced the gummi condition on "major change" (no definition, no decision maker who declares someting a "major change") with a clear rule on when sth is considered a "major change" - based on explicit core dev statement
* added timelines for decision making - 3 work days for non-major change, 12 work days for major change, 25 work days for stage 3 vote (if needed)
* clarifying what happens if lazy consensus is found (was not stated) - then PR must be merged by a core dev
* clarified that STEP must contain at least high-level design

### CODE REVIEW

**Comment by fkiraly** [comment]
from CC meeting on 2021-07-05:
FK to clarify approval on PR vs positive vote
CC members to read and review until next meeting

**Comment by mloning** [comment]
Hi all, here are a few thoughts on the suggested changes.

Regarding what happens if PRs do not receive any reply within a given amount of time: 

If a PR doesn't receive any replies, from my experience, it indicates that
* the proposed changes are perhaps not relevant at this point in time
* the changes have not been explained or communicated well
* there is something related to the PR makes people shy away from it

all of which for me would be indications to not merge the PR. 

Of course there may be a PR with none of the above problems that simply has not been commented on by anyone. In that case, to me the simpler solution would be to rely on core devs to communicate with each other and ask for reviews, rather than creating a rule where a PR gets merged automatically after certain amount of time.  

Regarding the right to have a review: I think we already cover this as a responsibility of the code owners, we can re-emphasise this, my main question about the right is how we would guarantee this and who would be responsible for guaranteeing this?

A few other suggestions: 
* rename "algorithm maintainers" to "code owners" (this is supported by GitHub and it's not just about algorithms)
* create speciality roles for different topics/modules who can help with triaging issues/PRs (thanks to @RNKuhns for the idea)

**Comment by fkiraly** [comment]
>If a PR doesn't receive any replies, from my experience, it indicates that
> there is something related to the PR makes people shy away from it

Indeed, and that, I think is a key point that I want to address!

What I want to avoid is not developing a key area due to social dynamics unrelated to the content, like avoidance where there is no "perfect" solution (perfect is the enemy of good!), apathy/inertia, or proctrastination. 
A similar topic is disagreement between core devs where onlookers think providing a review means "picking a side". Complete avoidance of review responsibilities should not be "the easy way out" from difficult design or content discussions, dynamics like that can easily drive away active core contributors too (by getting used to that avoidance stance).

None of these dynamics are, in general, the pull requester's fault, they contribute to sclerotic decision making culture, and hence are typically to the detriment of health and development of sktime.

**Comment by RNKuhns** [comment]
I personally think there are some things I think we could all do better as a group in terms of process and explanation. 

I'm willing to admit that given all the different developments with `sktime` that I will find myself unsure of what is being worked on, where I should focus my time (outside of functionality I'm working on adding) and most importantly how that fits into the larger plan. 

Many of us have a lot going on (e.g. full time jobs, completing degree programs, families, etc) and I think we should try and make it as easy as possible on everyone to focus their contribution time on substantive contribution (feedback/review, contributing issues/PRs, coordinating work, etc) as opposed to trying to stay caught up on everything that is going on. 

In terms of issues/PRs that are generated by the `sktime` development team, I'd like to see us have a more detailed roadmap of future development. This has a long-run component (in the long-run we'd like to have functionality X, Y, and Z) and a more tactical shorter-run component to help us coordinate our efforts better. This would include identifying areas of prioritized development in short-to-medium time horizons and having task boards for different things we'd like to add in different topic/module/functionality areas.

I think this helps because:

- It makes it much easier to understand what we plan to develop so developers can focus their efforts and therefore have more time for feedback/review, contributing issues/PRs, etc

- Authors of issues/PRs can link their explanations of the issues/PRs back to the plan for context and they can more easily focus on their explanation of what is most important for feedback and any important decision points they see related to the issue/PR


Note that I'm not advocating for us to rigidly just follow a set of task lists -- important and/or interesting things will pop-up along the way and will be worth working on in-lieu of what we planned on working on at that moment. The task lists and plans should be dynamic, we should just keep track of that dynamism.

Of course there will (hopefully) always be issues/PRs from the broader user community that will pop up. I think proposal on creating specialty roles for topics/modules (that was mentioned in @mloning's comment above) would help organize this process and reduce the burden on individual people (goal would be to try to have multiple people on topics/modules). The general idea on this front would be if an issue/PR comes in related to someone's topic/module area they'd engage on the PR provide initial triaging (is it actually a bug or is something else going on?, does the feature already exist or should it be added?, etc) and help shepherd along the issue/PR and make sure to get other core-developer input as needed.

**Comment by fkiraly** [comment]
>I think proposal on creating specialty roles for topics/modules (that was mentioned in @mloning's comment above) would help organize this process and reduce the burden on individual people (goal would be to try to have multiple people on topics/modules). The general idea on this front would be if an issue/PR comes in related to someone's topic/module area they'd engage on the PR provide initial triaging (is it actually a bug or is something else going on?, does the feature already exist or should it be added?, etc) and help shepherd along the issue/PR and make sure to get other core-developer input as needed.

Hm, @RNKuhns, that's not a bad idea.

I would consider this is a separate matter though, i.e., introduction of a new role, and a proposal how this relates to PR reviews.
Do you maybe want to draft something and PR to the governance document?

Also, it's worth thinking about how much there is to be codified as "laws", and how much is better left to improvement of our project planning practices, e.g., via the community team.

Further, have you thought about what happens if new modules come up, or how this proposal of module or topic owners would interact with improvements to the infrastructure of the entire package? Think about the tag logic, for instance.

**Comment by RNKuhns** [comment]
@fkiraly I'll work something up on the roles and talk with the community-team about it. I also agree that the first part is mainly about organization on all of our ends and that is also something the community-team can work on.

**Comment by TonyBagnall** [comment]
I'll split reply into two, first bits I think are not contentious 
> * making concrete conditions for core developers becoming active or inactive
agreed
> * making explicit that core developer approval for proposals must be by a core developer who is not the proposer
agreed

> * clarified where approvals/rejections must be put
"Votes are cast publicly, in the location of the proposed change"
I'm not sure what "in the location of the proposed change" means. Votes are not anonymous? Could we not run a poll instead?

> There's one other unrelated change to the sentence at the start, about welcoming people "who have experienced discrimination" in the past. I'm not sure it's a good idea do define groups by their suffering rather than self-stated labels or qualifiers, so I changed it to "groups underrepresented in the tech sector".
agreed.

**Comment by TonyBagnall** [comment]
the rest relates to PRs from core developers which are neither approved nor vetoed. On reflection, I am personally not in favour of the whole lazy approval mechanism as described. I would prefer rules along the lines of all PRs should receive a review within X days and if this does not happen the PR will automatically be put on the agenda for the next CC meeting or technical meeting where a review will be solicited

**Comment by mloning** [comment]
Re the lazy lazy consensus, I like @TonyBagnall suggestion: if a PR by a core dev hasn't received any replies within X days, it will be put on the agenda in the next technical discussion meeting, if that doesn't happen it is escalated to the next CC meeting.

**Comment by mloning** [comment]
My preferred solution: If a PR by a core developer has not received any replies within 10 working days, the author of the PR has the right to put the PR up for discussion as the top agenda item in the next technical committee.

**Comment by fkiraly** [comment]
>My preferred solution: If a PR by a core developer has not received any replies within 10 working days, the author of the PR has the right to put the PR up for discussion as the top agenda item in the next technical committee.

Ok, that's what we agreed on - more specifically, it goes at the top of the agenda for *decision*.
Also, we don't have a technical committee yet, I assume it's the regular core dev meeting, right?

File: GOVERNANCE.rst [hunk 1]
```diff
@@ -335,16 +346,51 @@ sktime uses a “consensus seeking” process for making decisions. The
...
 core developers.
 
--  To accept proposed changes, we require approval by one core developer
-   (lazy consensus) and no rejection by a core developer (veto right).
+-  For a proposed changes to be approved via lazy consensus, one of the following
+   two conditions must be satisfied:
+    - approval by at least one core developer and no rejection by a core developer (veto right).
+      The approval required for this condition must be by a core developer different from a proposer.
+    - approval in the regular core developer meeting as part of the agenda.
+      This can happen only if no approvals or rejections are expresesd on a PR, see below.
```
Comment (mloning) [inline GOVERNANCE.rst:354]: Let's rely on explicit approvals on GitHub
```suggestion
```
Also makes it easier: you need one approval (lazy consensus) and no rejection (veto right)

File: GOVERNANCE.rst [hunk 2]
```diff
@@ -335,16 +346,51 @@ sktime uses a “consensus seeking” process for making decisions. The
...
+-  For a proposed change to be rejected via lazy consensus, one of the following two conditions
+   must be satisfied:
+    - rejection by at least one core developer, and no acceptance by a core developer.
+    - rejection in the regular core developer meeting as part of the agenda.
+      This can happen only if no approvals or rejections are expresesd on a PR, see below.
```
Comment (mloning) [inline GOVERNANCE.rst:359]:
```suggestion
```

File: GOVERNANCE.rst [hunk 3]
```diff
@@ -384,24 +431,32 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
...
+For all stage 3 decisions of the CC on major changes,
+an sktime enhancement proposal (STEP) must have been made
 public and discussed before the vote.
 
 An sktime enhancement proposal is a consolidated document, with a concise
 problem statement, a clear description of the proposed solution and a
 comparison with alternative solutions, as outlined in our
 `template <https://github.com/sktime/enhancement-proposals/blob/master/TEMPLATE.md>`__.
 
+A complete STEP must always include at least a high-level design for the proposed change,
```
Comment (mloning) [inline GOVERNANCE.rst:457]: I think more info on what consistute a (good) STEP should go into the README of the enhancement-proposal repo.

File: GOVERNANCE.rst [hunk 4]
```diff
@@ -335,16 +346,51 @@ sktime uses a “consensus seeking” process for making decisions. The
...
+   respectively, or any written comments
+   containing "I formally approve/reject" in reference to the proposal.
+   If the proposed change has the form of a PR, approval must be in the form of a GitHub PR approval.
+-  Major changes to the project require a more
    detailed analysis and a consensus that is both explicit and informed.
-   These changes require an :ref:`steps`.
+   A major change always require a worked out :ref:`steps`.
+   A change is major in the above sense
+   if at least three core developers explicitly state that the proposal is a major change
+   and requires especial consideration, in the location.
```
Comment (mloning) [inline GOVERNANCE.rst:373]: I think some changes are major changes even if less than three core devs explicitly state that on the PR/issue. I'd say we leave out the definition of what counts as "major" here and resolve that through discussion if unclear. 
```suggestion
```

File: GOVERNANCE.rst [hunk 5]
```diff
@@ -335,16 +346,51 @@ sktime uses a “consensus seeking” process for making decisions. The
...
+-  Proposers are expected to give reasonable time for consideration, that is,
+   time and opportunity for core developers to review and
+   give their opinion on the contribution.
```
Comment (mloning) [inline GOVERNANCE.rst:376]:
```suggestion
   give their opinion on the pull request.
```

File: GOVERNANCE.rst [hunk 6]
```diff
@@ -335,16 +346,51 @@ sktime uses a “consensus seeking” process for making decisions. The
...
+   Ten working days excluding week-ends constitute "reasonable time" in the above sense.
+   For pull requests, the period resets at every new change made to the PR.
+   It starts only when all GitHub checks pass.
+-  If the "reasonable time" period elapses and no approval or rejection has been expressed on a PR,
+   the PR is scheduled for decision, at the top of the next regular core developer meeting's agenda
+   in which less than seven PR are already on the agenda. If no regular core developer meeting
+   is scheduled within the two calendar weeks after the period elapses,
+   the core developers must schedule one within those two weeks.
```
Comment (mloning) [inline GOVERNANCE.rst:384]: I'd rephrase this as in my comment, since this only applies to PRs by core devs 
```suggestion
-  If a PR by a core developer has not received any approvals or rejections within ten working days excluding weekends, the core developer who opened the PR has the right to put the PR up for discussion at the top agenda in the next regular core developer meeting and receive a review from at least one of the other core developers. 
```

File: GOVERNANCE.rst [hunk 7]
```diff
@@ -335,16 +346,51 @@ sktime uses a “consensus seeking” process for making decisions. The
...
+-  Decision, in the core developer meeting, is by 2/3 majority of core developers in the meeting,
+   possibly including the proposer.
+   Decision is between approval or rejection. Approval can be subjected to minor corrections, which
+   requires an additional, simple majority consensus decision of core developers in the meeting.
+
+Approval or rejection as above count as "consensus".
+Lack of such consensus can arise under two conditions:
+- at least one approval and at least one rejection on GitHub
+- no 2/3 majority for either approval or rejection in the core developer meeting
```
Comment (mloning) [inline GOVERNANCE.rst:393]:
```suggestion
```

Comment (mloning) [reply GOVERNANCE.rst:393]: For me, the main purpose of adding it to the agenda is for the PR to receive at least some feedback, at this stage we don't have to vote yet, voting is also already handled in stage 2, so I'd remove these changes here.

File: GOVERNANCE.rst [hunk 8]
```diff
@@ -357,11 +403,12 @@ When no consensus can be found, any core developer can call for a vote
 at any point during the discussion.
 
 -  The vote will conclude 5 days excluding weekends from the call for the vote.
--  Votes are voluntary. Abstentions are allowed. You can
+-  Votes are voluntary. Abstentions are allowed. Core developers can
    abstain by simply not casting a vote.
 -  All votes are a binary vote: for or against accepting the proposed
    changes.
 -  Votes are casts as comments: +1 (approval) or -1 (rejection).
+-  Votes are cast publicly, in the location of the proposed change.
```
Comment (mloning) [inline GOVERNANCE.rst:411]:
```suggestion
-  Votes are cast publicly, on the pull request.
```

File: GOVERNANCE.rst [hunk 9]
```diff
@@ -384,24 +431,32 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
 the CC tries to resolve the deadlock.
 
 -  The CC will use consensus seeking.
--  If no consensus can be found within a month, the decision is made
-   through a simple majority vote (with tie breaking) among the CC
+-  If no consensus can be found within twenty working days excluding week-ends
```
Comment (mloning) [inline GOVERNANCE.rst:434]:
```suggestion
-  If no consensus can be found within twenty working days excluding weekends
```

**Review by mloning — COMMENTED** [review]
Hi @fkiraly, when you look at how the change render, the formatting looks a bit off, perhaps just a spacing issue, the rst format is sometimes a bit tricky.
 
A few comments below. Happy to help with these changes if you like.

File: GOVERNANCE.rst (hunk 1 above)
Comment (fkiraly) [reply GOVERNANCE.rst:354]: but that's still an option! I'm just spelling out the approval conditions. Either what you say (approval, no veto), or approval in the core dev meeting if neither approval nor veto appears on the PR.

File: GOVERNANCE.rst (hunk 2 above)
Comment (fkiraly) [reply GOVERNANCE.rst:359]: no? I thought we agreed that if no approvals/rejections appear, it is tabled at the top of agenda at the core dev meeting.

File: GOVERNANCE.rst (hunk 3 above)
Comment (fkiraly) [reply GOVERNANCE.rst:457]: yes, but that's additional work, right? Should we open an issue?

File: GOVERNANCE.rst (hunk 4 above)
Comment (fkiraly) [reply GOVERNANCE.rst:373]: well, but if we don't define what is "major", we cannot make "major" a condition in the rules, since it is undefined and arbitrary otherwise. It appears as a condition for PR that require a STEP.

So either we remove both the requirement of STEPS too, or we leave both things in.

File: GOVERNANCE.rst (hunk 5 above)
Comment (fkiraly) [reply GOVERNANCE.rst:376]: not all contributions are pull requests - or do we assume they are and must be?

File: GOVERNANCE.rst (hunk 6 above)
Comment (fkiraly) [reply GOVERNANCE.rst:384]: this would make the process less clear:

* how do we prevent too long core dev meetings since there are so many PR?
* how do we prevent that no core dev meeting is planned, is it then indefinitely stalled?
* what does "review" mean? If it is a GitHub review, why is a core dev meeting needed? As opposed to a reviewer being appointed with the duty to review?

File: GOVERNANCE.rst (hunk 7 above)
Comment (fkiraly) [reply GOVERNANCE.rst:393]: Not against this, but please note that by simply removing this, the process is incomplete. How does it reach stage 2 in your preferred process? There is no link between moving it to the core dev meeting and stage 2.

File: GOVERNANCE.rst (hunk 8 above)
Comment (fkiraly) [reply GOVERNANCE.rst:411]: are all proposals pull requests? See above.

File: GOVERNANCE.rst (hunk 9 above)
Comment (fkiraly) [reply GOVERNANCE.rst:434]: appears I'm "old-fashioned", might have been reading too many old books.

A poster puts it nicely [here on quora](https://www.quora.com/Which-is-correct-weekend-or-week-end)

"
Weekend is in 2021 (and has been for many decades) the dominant form.

If you were to use week-end in writing, you would probably be labeled as old-fashioned or as fond of outdated language.

Still, there was a time when weekend was inconceivable, a time when it would have been almost obscene.

To write, around the 1940s:

Where have you been for the weekend?

Would've been just like saying, in 2021:

Who will be the garbage taker outer today?

(Which, by the way, you can translate into English by saying: Who will take out the garbage today?)

Probably one day such an expression will become standard English too.

Back to your question, the turning point was somewhere in the fifties, when weekend was introduced and gradually gained popularity over week-end.

Again, before this time week-end was the only form.
"

**Comment by fkiraly** [comment]
from discussion in 2021-08-09 CC meeting: Pr should not be for approval in core dev meeting, but for discussion & assigning review

File: GOVERNANCE.rst (hunk 5 above)
Comment (fkiraly) [reply GOVERNANCE.rst]: changed all contribs to PR

File: GOVERNANCE.rst (hunk 6 above)
Comment (fkiraly) [reply GOVERNANCE.rst]: clarified that the dev meeting appoints a reviewer.

File: GOVERNANCE.rst (hunk 7 above)
Comment (fkiraly) [reply GOVERNANCE.rst]: simplified

File: GOVERNANCE.rst (hunk 4 above)
Comment (fkiraly) [reply GOVERNANCE.rst]: deferred to the table

File: GOVERNANCE.rst (hunk 2 above)
Comment (fkiraly) [reply GOVERNANCE.rst]: simplified

**Comment by fkiraly** [comment]
I've addressed the comments by simplifying the process:

* core dev meeting appoints a reviewer who needs to review and approve/reject
* there is only one condition for lazy consensus on approve or reject, so the process greatly simplifies
* "major change" definition is deferred to the table in the section before (which already was there, it is unchanged from pre-PR state); the concept of "major change" is removed, instead the table simply says when a STEP is needed

Due to the last point, I introduced a minor change:

* since all API changes would require STEPs strictly speaking by the table, I've made a modification which allows a STEP to be a short summary in an issue or PR if it is small; the CC decides when a longer STEP is needed.

File: GOVERNANCE.rst [hunk 10]
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
...
    * - Changes to sktime's governance (this document and the CoC)
-     - Lazy consensus based on an :ref:`steps`
+     - Lazy consensus, requires a :ref:`steps`
    * - Appointment
-     - Voting
+     - Directly starts with voting (stage 2)
 
 .. _stage-1:
 
-Stage 1: lazy consensus with veto right
```
Comment (mloning) [inline GOVERNANCE.rst:331]: I prefer the previous header since that's what it is

File: GOVERNANCE.rst [hunk 11]
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
...
-   give their opinion on the pull request if they’re not confident
-   others would agree.
--  More important changes that impact the full project require a more
-   detailed analysis and a consensus that is both explicit and informed.
-   These changes require an :ref:`steps`.
+-  Proposed changes should be in the form of GitHub pull requests (PR).
+   Some changes also require a worked out :ref:`steps`, this depends on the type of change, see
+   `decision making process <#Decision-making>`__ above.
+-  For a proposed change to be approved via lazy consensus, it needs to
+   approval by at least one core developer and no rejection by a core developer (veto right).
```
Comment (mloning) [inline GOVERNANCE.rst:353]: to explain the header 
```suggestion
   approval by at least one core developer (lazy consensus) and no rejection by a core developer (veto right).
```

File: GOVERNANCE.rst [hunk 12]
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
...
--  Approvals and rejections can be expressed as +1 and -1 comments,
-   respectively.
--  Core developers are expected to give reasonable time to others to
-   give their opinion on the pull request if they’re not confident
-   others would agree.
--  More important changes that impact the full project require a more
-   detailed analysis and a consensus that is both explicit and informed.
-   These changes require an :ref:`steps`.
+-  Proposed changes should be in the form of GitHub pull requests (PR).
+   Some changes also require a worked out :ref:`steps`, this depends on the type of change, see
```
Comment (mloning) [inline GOVERNANCE.rst:350]:
```suggestion
   Some changes also require a worked out :ref:`steps`. This depends on the type of change, see
```

File: GOVERNANCE.rst [hunk 13]
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
...
+   `decision making process <#Decision-making>`__ above.
+-  For a proposed change to be approved via lazy consensus, it needs to
+   approval by at least one core developer and no rejection by a core developer (veto right).
+   The approval required for this condition must be by a core developer different from a proposer.
```
Comment (mloning) [inline GOVERNANCE.rst:354]:
```suggestion
   Core developer cannot approve their proposed PRs. 
```

File: GOVERNANCE.rst [hunk 14]
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
...
+-  For a proposed change to be rejected via lazy consensus, it needs to receive a
+   rejection by at least one core developer, and no acceptance by a core developer.
+-  Approvals must be in the form of a GitHub PR approval. of the PR in question.
```
Comment (mloning) [inline GOVERNANCE.rst:357]:
```suggestion
-  Approvals must be in the form of a GitHub PR approval of the PR in question.
```

File: GOVERNANCE.rst [hunk 15]
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
...
+   Rejections can be expressed as -1 comments, or any written comments
+   containing "I formally reject" in the PR, in reference to it.
```
Comment (mloning) [inline GOVERNANCE.rst:359]:
```suggestion
   Rejections must be in the form of a GitHub PR rejection on the PR in question.
```

File: GOVERNANCE.rst [hunk 16]
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
...
+-  Proposers are expected to give reasonable time for consideration, that is,
+   time and opportunity for core developers to review and
+   give their opinion on the PR.
+   Ten working days excluding week-ends constitute "reasonable time" in the above sense.
+   The period resets at every new change made to the PR.
+   It starts only when all GitHub checks pass.
+-  During this period, the PR can be merged if it has an approval and no rejection, but must be
+   reverted if it receives a rejection in addition.
```
Comment (mloning) [inline GOVERNANCE.rst:367]: I think this isn't practical and will introduce more confusion. 
```suggestion
-  During this period, the PR can be merged if it has an approval and no rejection, but must be
   reverted if it receives a rejection in addition.
```

File: GOVERNANCE.rst [hunk 17]
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
...
+-  If the "reasonable time" period elapses and no approval or rejection has been expressed on a PR,
+   the PR is scheduled at the top of the next regular (bi-monthly) core developer meeting's agenda.
```
Comment (mloning) [inline GOVERNANCE.rst:369]:
```suggestion
   the PR is scheduled at the top of agenda for the next developer meetup.
```
We should add a link to the meet up page on the new website here

File: GOVERNANCE.rst [hunk 18]
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
...
+   In that meeting, a core developer is assigned to review, and approve or reject within
+   5 days of the meeting, excluding week-ends.
```
Comment (mloning) [inline GOVERNANCE.rst:371]:
```suggestion
   In that meeting, a core developer is assigned to review the PR and either approve or reject within five days of the meeting excluding weekends.
```

File: GOVERNANCE.rst [hunk 19]
```diff
@@ -320,31 +331,47 @@ corresponding decision making process is described in more detail below.
...
+
+Failure of lazy consensus, in the above sense, can arise only under the following condition:
+at least one approval and at least one rejection in the PR.
```
Comment (mloning) [inline GOVERNANCE.rst:374]:
```suggestion
Failure of lazy consensus, in the above sense, can arise if a PR has received at least one approval and at least one rejection.
```

File: GOVERNANCE.rst [hunk 20]
```diff
@@ -353,20 +380,23 @@ When no consensus can be found, the decision is escaled to :ref:`stage-2`.
 Stage 2: voting
 ~~~~~~~~~~~~~~~
 
-When no consensus can be found, any core developer can call for a vote
-at any point during the discussion.
+Voting takes place:
+
+* when no lazy consensus can be found in stage 1 decision making
```
Comment (mloning) [inline GOVERNANCE.rst:385]:
```suggestion
* when no lazy consensus can be found in stage 1 above
```

File: GOVERNANCE.rst [hunk 21]
```diff
@@ -384,27 +414,41 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
 the CC tries to resolve the deadlock.
 
 -  The CC will use consensus seeking.
--  If no consensus can be found within a month, the decision is made
-   through a simple majority vote (with tie breaking) among the CC
+-  If no consensus can be found within twenty working days excluding weekends
+   since the beginning of the stage-1 "reasonable time for consideration" period,
+   the decision is made through a simple majority vote (with tie breaking) among the CC
```
Comment (mloning) [inline GOVERNANCE.rst:419]:
```suggestion
-  If no consensus can be found, the decision is made through a simple majority vote (with tie breaking) among the CC
```

Comment (mloning) [reply GOVERNANCE.rst:419]: there may not always be a CC meeting within that period, the role of the CC meeting in that case would also be to try again to find a consensus, by perhaps taking parts of the PR and rejecting other, etc

File: GOVERNANCE.rst [hunk 22]
```diff
@@ -384,27 +414,41 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
...
    members.
--  Any CC decision must be supported by an :ref:`steps`, which has been made public and discussed before the vote.
+-  If this vote is needed, the vote should take place within twenty-five working days after the beginning
+   of the stage-1 "reasonable time for consideration" period, during a regular, minuted CC meeting.
+-  Any proposal reaching stage 3 must be supported by an :ref:`steps`,
+   which has been made public at least 5 days, excluding week-ends, before the vote.
```
Comment (mloning) [inline GOVERNANCE.rst:424]:
```suggestion
   which has been made public at least 5 days, excluding weekends, before the vote.
```

File: GOVERNANCE.rst [hunk 23]
```diff
@@ -384,27 +414,41 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
...
 -  The CC will use consensus seeking.
--  If no consensus can be found within a month, the decision is made
-   through a simple majority vote (with tie breaking) among the CC
+-  If no consensus can be found within twenty working days excluding weekends
+   since the beginning of the stage-1 "reasonable time for consideration" period,
+   the decision is made through a simple majority vote (with tie breaking) among the CC
    members.
--  Any CC decision must be supported by an :ref:`steps`, which has been made public and discussed before the vote.
+-  If this vote is needed, the vote should take place within twenty-five working days after the beginning
+   of the stage-1 "reasonable time for consideration" period, during a regular, minuted CC meeting.
```
Comment (mloning) [inline GOVERNANCE.rst:422]:
```suggestion
```

File: GOVERNANCE.rst [hunk 24]
```diff
@@ -384,27 +414,41 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
...
+   which has been made public at least 5 days, excluding week-ends, before the vote.
 
 .. _steps:
 
 sktime enhancement proposal
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-For all decision of the CC, an sktime enhancement proposal (STEP) must have been made
-public and discussed before the vote.
+sktime enhancmenet proposals (STEPs) are required for:
```
Comment (mloning) [inline GOVERNANCE.rst:431]:
```suggestion
sktime enhancement proposals (STEPs) are required for:
```

File: GOVERNANCE.rst [hunk 25]
```diff
@@ -384,27 +414,41 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
...
+
+If a STEP is required by a vote, it must have been made public at least 5 working days (excluding week-ends) before that vote.
+
+A STEP is a consolidated document, with a concise
 problem statement, a clear description of the proposed solution and a
 comparison with alternative solutions, as outlined in our
 `template <https://github.com/sktime/enhancement-proposals/blob/master/TEMPLATE.md>`__.
 
-We collect and discuss proposals in sktime’s `repository for
+A complete STEP must always include at least a high-level design for the proposed change,
```
Comment (mloning) [inline GOVERNANCE.rst:443]: As before, let's define STEPs in a single place in the README in the STEPS repo, so a link should do here.

File: GOVERNANCE.rst [hunk 26]
```diff
@@ -384,27 +414,41 @@ If the proposed change cannot gather a 2/3 majority of the votes cast,
...
+not just a wishlist of features.
+
+Usually, we collect and discuss proposals in sktime’s `repository for
 enhancement-proposals <https://github.com/sktime/enhancement-proposals>`__.
 
+For smaller changes, such as punctual changes to the API or governance documents,
+the STEP can also be be part of an issue or pull request.
```
Comment (mloning) [inline GOVERNANCE.rst:450]: for smaller changes we don't need a STEP I think, every PR should have some description of what's going on, but that's not what we consider a STEP I'd say 
```suggestion
For smaller changes, such as punctual changes to the API or governance documents,
the STEP can also be be part of an issue or pull request.
```

File: GOVERNANCE.rst (hunk 23 above)
Comment (fkiraly) [reply GOVERNANCE.rst:422]: why would you want to remove this very generous time limit? It would introduce a problematic loophole where decision making can stall simply since the cc never votes on the PR.

We want to avoid dynamics where indefinite stalling of a decision is a viable option.

File: GOVERNANCE.rst (hunk 26 above)
Comment (fkiraly) [reply GOVERNANCE.rst:450]: then this would be in contradiction with the pre-existing requirements for decision making, which require a STEP for every change to the API or the governance document.

File: GOVERNANCE.rst (hunk 13 above)
Comment (fkiraly) [reply GOVERNANCE.rst:354]: this suggestion would not be clear in a case where multiple people work on a PR?

File: GOVERNANCE.rst (hunk 15 above)
Comment (fkiraly) [reply GOVERNANCE.rst:359]: there is no such thing as a GitHub PR rejection?

File: GOVERNANCE.rst (hunk 16 above)
Comment (fkiraly) [reply GOVERNANCE.rst:367]: well, what's your alternative process? We must always wait for the period to elapse to merge? I don't think we have done this for a single PR.

The other alternative would leave a loophole where proposer and reviewer collude, the reviewer merging a few seconds after a PR appears; or, the less extreme but still problematic version of this where some core devs do not have enough time to review, e.g., since the PR happens on a weekend.

File: GOVERNANCE.rst (hunk 25 above)
Comment (fkiraly) [reply GOVERNANCE.rst:443]: that would make that readme a formal document and an appendix to this one, are you sure?

File: GOVERNANCE.rst (hunk 19 above)
Comment (fkiraly) [reply GOVERNANCE.rst:373]: it's not an "if" (sufficiency) but an "only if" (necessity), see here for terminology:
https://en.wikipedia.org/wiki/Necessary_and_sufficient_conditions#Definitions

(also, the "can" would be wrong in case you mean sufficiency)

File: GOVERNANCE.rst (hunk 26 above)
Comment (mloning) [reply GOVERNANCE.rst:449]: Yes every "bigger" change

File: GOVERNANCE.rst (hunk 25 above)
Comment (mloning) [reply GOVERNANCE.rst:442]: As said before, I don't see this as a legal document and I don't think it should be, so not really sure what "formal document" or "appendix" means, having a link should do the job

File: GOVERNANCE.rst (hunk 10 above)
Comment (fkiraly) [reply GOVERNANCE.rst:331]: ok

File: GOVERNANCE.rst (hunk 26 above)
Comment (fkiraly) [reply GOVERNANCE.rst:449]: it doesn't say that in the table (section decision making)

File: GOVERNANCE.rst (hunk 25 above)
Comment (mloning) [reply GOVERNANCE.rst:442]: I suggest we have the STEP guidelines on the website at: https://www.sktime.org/en/latest/enhancement_proposals.html

Comment (mloning) [reply GOVERNANCE.rst:442]: So we can easily link it there

File: GOVERNANCE.rst (hunk 26 above)
Comment (mloning) [reply GOVERNANCE.rst:449]: Sure, but I wouldn't call PR descriptions STEPS, so I'd say for smaller changes no STEP is needed

File: GOVERNANCE.rst (hunk 13 above)
Comment (mloning) [reply GOVERNANCE.rst:354]: if multiple core devs work on a PR and review each other code, that's fine no? of course it's even better if other core devs review too, especially if it's bigger API changes etc

File: GOVERNANCE.rst (hunk 23 above)
Comment (mloning) [reply GOVERNANCE.rst:421]: I know, but it just adds to many conditions/rules/timescales which are hard to keep track of, the CC will be working to resolve conflicts, don't think we need to impose more time limits here

File: GOVERNANCE.rst (hunk 13 above)
Comment (fkiraly) [reply GOVERNANCE.rst:354]: no, it's *not* fine in my opinion. You can't review your own stuff, it's not a review.

File: GOVERNANCE.rst (hunk 26 above)
Comment (fkiraly) [reply GOVERNANCE.rst:449]: ok - suggestion is identical to the original, so I assume this is resolved.

File: GOVERNANCE.rst (hunk 13 above)
Comment (fkiraly) [reply GOVERNANCE.rst:354]: let's discuss separately whether we allow PR authors to review their own PR, we can discuss at the next CC meeting. Until then, let's keep the status quo here.

File: GOVERNANCE.rst (hunk 23 above)
Comment (fkiraly) [reply GOVERNANCE.rst:421]: ok, fine. Let's discuss this addition separately then, I still think giving timelines is important.

File: GOVERNANCE.rst (hunk 25 above)
Comment (fkiraly) [reply GOVERNANCE.rst:440]: yes, this is already linked, so I assume this is resolved.

File: GOVERNANCE.rst (hunk 21 above)
Comment (fkiraly) [reply GOVERNANCE.rst:418]: this would muddy the definition of what it means "no consensus can be found". When do we pass to the CC?

I really don't like that your suggested changes are reducing accountability and guarantee of speedy process. Whether you intend it or not, removal of clear condtions and clear decision making points is directly hampering effective governance.

File: GOVERNANCE.rst (hunk 16 above)
Comment (fkiraly) [reply GOVERNANCE.rst:367]: The suggested change does not change anything, so I assume you are fine with the content.

**Comment by fkiraly** [comment]
I have now updated this with what I think was the remaining point outstanding (add the "veto right" explicitly back in the "lazy consensus" section header).

I've also moved the document to the new location, and updated the links where they were pointing to the roles list.
